core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -34,64 +35,66 @@
     }
 
     public static ParseState createFromByteStream(final ByteStream input) {
         return createFromByteStream(input, ZERO);
     }
 
     public ParseState addBranch(final Token token) {
-        return new ParseState(order.addBranch(token), source, offset, token.isIterable() ? iterations.add(new ImmutablePair<>(token, ZERO)) : iterations, references);
+        return new ParseState(order.addBranch(token), cache, source, offset, token.isIterable() ? iterations.add(new ImmutablePair<>(token, ZERO)) : iterations, references);
     }
 
     public ParseState closeBranch(final Token token) {
         if (token.isIterable() && !iterations.head.left.equals(token)) {
             throw new IllegalStateException(format("Cannot close branch for iterable token %s. Current iteration state is for token %s.", token.name, iterations.head.left.name));
         }
-        return new ParseState(order.closeBranch(), source, offset, token.isIterable() ? iterations.tail : iterations, references);
+        return new ParseState(order.closeBranch(), cache, source, offset, token.isIterable() ? iterations.tail : iterations, references);
     }
 
     public ParseState add(final ParseReference parseReference) {
-        return new ParseState(order, source, offset, iterations, references.add(parseReference));
+        return new ParseState(order, cache, source, offset, iterations, references.add(parseReference));
     }
 
     public ParseState add(final ParseValue parseValue) {
-        return new ParseState(order.add(parseValue), source, offset, iterations, references);
+        return new ParseState(order.add(parseValue), cache.add(parseValue), source, offset, iterations, references);
     }
 
     public ParseState createCycle(final ParseReference parseReference) {
-        return new ParseState(order.add(parseReference), source, offset, iterations, references);
+        return new ParseState(order.add(parseReference), cache, source, offset, iterations, references);
     }
 
     public ParseState iterate() {
-        return new ParseState(order, source, offset, iterations.tail.add(new ImmutablePair<>(iterations.head.left, iterations.head.right.add(ONE))), references);
+        return new ParseState(order, cache, source, offset, iterations.tail.add(new ImmutablePair<>(iterations.head.left, iterations.head.right.add(ONE))), references);
     }
 
     public Optional<ParseState> seek(final BigInteger newOffset) {
-        return newOffset.compareTo(ZERO) >= 0 ? Optional.of(new ParseState(order, source, newOffset, iterations, references)) : Optional.empty();
+        return newOffset.compareTo(ZERO) >= 0 ? Optional.of(new ParseState(order, cache, source, newOffset, iterations, references)) : Optional.empty();
     }
 
     public ParseState withOrder(final ParseGraph order) {
-        return new ParseState(order, source, offset, iterations, references);
+        return new ParseState(order, NO_CACHE, source, offset, iterations, references);
```

Comment:
`withOrder` is used when using scopes. Since scopes define a limitation to the search region, which we do not track within the cache, so we should disable it. Note that it is only disabled when evaluating the expression.